notebooks/byod.py

```diff
@@ -72,134 +98,136 @@ def upload_columns(table: str, columns: Dict[str, List[Any]]):
+    def parse_cram_crai(filename: str):
+        if filename.endswith(".cram.crai"):
+            # double extension in the crai file, ie "foo.cram.crai"
+            sample = filename[:-10]
         else:
-            return set()
-
-    def join_keyed_rows(keyed_rows_a: Dict[str, Any], keyed_rows_b: Dict[str, Any]) -> Dict[str, Any]:
-        a_columns = keyed_row_columns(keyed_rows_a)
-        b_columns = keyed_row_columns(keyed_rows_b)
-        assert not a_columns.intersection(b_columns), "Keyed rows to join may not share columns"
-        common_keys = set(keyed_rows_a.keys()).union(set(keyed_rows_b.keys()))
-        return {k: dict(**keyed_rows_a.get(k, {c: BLANK_CELL_VALUE for c in a_columns}),
-                        **keyed_rows_b.get(k, {c: BLANK_CELL_VALUE for c in b_columns}))
-                for k in common_keys}
-
-    def join_data_tables(new_table: str, tables_to_join: list, join_column: str):
-        keyed_rows = get_keyed_rows(tables_to_join[0], join_column)
-        for table_name in tables_to_join[1:]:
-            keyed_rows = join_keyed_rows(keyed_rows, get_keyed_rows(table_name, join_column))
-        upload_rows(new_table, [{join_column: k, **row} for k, row in keyed_rows.items()])
+            sample = filename[:-5]
+        ext = filename[-4:]
+        return sample, ext
 
     def create_cram_crai_table(table: str, listing: Iterable[str]):
-        crams = dict()
-        crais = dict()
+        matched_files = defaultdict(dict)  # type: dict
+        allCrams = []
+        allCrais = []
         for key in listing:
             _, filename = key.rsplit("/", 1)
-
-            parts = filename.split(".")
-            if 3 == len(parts):  # foo.cram.crai branch
-                sample, _, ext = parts
-            elif 2 == len(parts):  # "foo.cram" or "foo.crai" branch
-                sample, ext = parts
-            else:
-                raise ValueError(f"Unable to parse '{filename}'")
-
-            if "cram" == ext:
-                crams[sample] = key
-            elif "crai" == ext:
-                crais[sample] = key
-            else:
-                continue
-        samples = sorted(crams.keys())
-        upload_columns(table, dict(sample=samples,
-                                   cram=[crams[s] for s in samples],
-                                   crai=[crais[s] for s in samples]))
+            sample, ext = parse_cram_crai(filename)
+            if ext in ['cram', 'crai']:
+                matched_files[sample][ext] = filename
+                if ext == 'cram':
+                    allCrams.append(sample)
+                else:
+                    allCrais.append(sample)
+
+        if allCrams != allCrais:
+            # We don't want to iterate an additional time if we don't need to,
+            # so this only runs if there's more crams than crais or vice versa
+            for sample in list(matched_files):
+                try:
+                    if not matched_files[sample]['cram']:
```

Comment:
Have you heard of matched_files[sample].get('cram')?  I think it might be easier and cleaner than the `try` / `except` logic throughout this.

Reply:
If I understand what you're suggesting to be...
```
    if allCrams != allCrais:
        # We don't want to iterate an additional time if we don't need to,
        # so this only runs if there's more crams than crais or vice versa
        for sample in list(matched_files):
            if not matched_files[sample].get('cram'):
                continue
            print(f'{sample} is missing a cram file and will be removed')
            del matched_files[sample]
            break
            if not matched_files[sample].get('crai'):
                continue
            print(f'{sample} is missing a crai file and will be removed')
            del matched_files[sample]
```
...then that won't work, it'll appear to pass but then throw a KeyError when creating the table to be uploaded. Neither does this, although I'm not quite sure why, seeing as it keeps the try/except blocks.

```
    if allCrams != allCrais:
        # We don't want to iterate an additional time if we don't need to,
        # so this only runs if there's more crams than crais or vice versa
        for sample in list(matched_files):
            try:
                if not matched_files[sample].get('cram'):
                    continue
            except KeyError:
                print(f'{sample} is missing a cram file and will be removed')
                del matched_files[sample]
                break
            try:
                if not matched_files[sample].get('crai'):
                    continue
            except KeyError:
                print(f'{sample} is missing a crai file and will be removed')
                del matched_files[sample]
```

Reply:
It would actually be something like this:

```
    if allCrams != allCrais:
        # We don't want to iterate an additional time if we don't need to,
        # so this only runs if there's more crams than crais or vice versa
        for sample in list(matched_files):
            if not (matched_files[sample].get('cram') or matched_files[sample].get('crai')):
                del matched_files[sample]
```

Reply:
That throws a KeyError later down the line.

Reply:
This:

```
    if allCrams != allCrais:
        # We don't want to iterate an additional time if we don't need to,
        # so this only runs if there's more crams than crais or vice versa
        for sample in list(matched_files):
            try:
                if not matched_files[sample]['cram']:
                    continue
            except KeyError:
                print(f'{sample} is missing a cram file and will be removed')
                del matched_files[sample]
                break
            try:
                if not matched_files[sample]['crai']:
                    continue
            except KeyError:
                print(f'{sample} is missing a crai file and will be removed')
                del matched_files[sample]

        crams = []
        crais = []
        samples = []
        for sample in matched_files:
            samples.append(sample)
            for ext in ('cram', 'crai'):
                if ext == 'cram':
                    crams.append(f"{bucket}/{subdirectory}/{matched_files[sample][ext]}")
                else:
                    crais.append(f"{bucket}/{subdirectory}/{matched_files[sample][ext]}")
```

Seems to be the same as this:

```
for sample in list(matched_files):
    if matched_files[sample].get('cram') and matched_files[sample].get('crai'):
        crams.append(f"{bucket}/{subdirectory}/{matched_files[sample]['cram']}")
        crais.append(f"{bucket}/{subdirectory}/{matched_files[sample]['crai']}")
    else:
        print(f'{sample} is missing a file and will be removed:\n{json.dumps(matched_files[sample])}')
```